api/metrics/label_gatherer.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package metrics
+
+import (
+	"cmp"
+	"errors"
+	"fmt"
+	"slices"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+var (
+	_ MultiGatherer = (*prefixGatherer)(nil)
+
+	errDuplicateGatherer = errors.New("attempt to register duplicate gatherer")
+)
+
+// NewLabelGatherer returns a new MultiGatherer that merges metrics by adding a
+// new label.
+func NewLabelGatherer(labelName string) MultiGatherer {
+	return &labelGatherer{
+		labelName: labelName,
+	}
+}
+
+type labelGatherer struct {
+	multiGatherer
+
+	labelName string
+}
+
+func (g *labelGatherer) Register(labelValue string, gatherer prometheus.Gatherer) error {
+	g.lock.Lock()
+	defer g.lock.Unlock()
+
+	if slices.Contains(g.names, labelValue) {
```

Comment:
> Why is it desirable to throw an error vs silently accepting?

Silently accepting will result in silently dropping metrics because of a bug in the metrics initialization. I think it's better for us to clearly see that metrics aren't being reported correctly rather than needing to discover that metrics aren't showing up and then needing to investigate.

> why a slice instead of a set

`g.names` is inherited by `multiGatherer`. For the implementation of `prefix_gatherer` we need to compare against all the strings, so I didn't feel like using a `set` made much sense.